utils/big_decimal.cc

```diff
@@ -143,6 +143,97 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
     return x.compare(y) <=> 0;
 }
 
+std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+{
+    if (_scale == other._scale) {
+        return _unscaled_value.compare(other._unscaled_value) <=> 0;
+    }
+
+    // boost::multiprecision::sign() returns -1, 0 or 1
+    const int sign = boost::multiprecision::sign(_unscaled_value);
+    const int sign_other = boost::multiprecision::sign(other._unscaled_value);
+    if (sign != sign_other) {
+        return sign <=> sign_other;
+    }
+    // At this point we know the two signs are equal, so if sign == 0, both signs
+    // and consequently both numbers are zeros.
+    if (sign == 0) {
+        return std::strong_ordering::equal;
+    }
+
+    // At this point we know that both numbers have the same sign, so if one is negative, the other is too.
+    const bool comparing_negatives = (sign < 0);
+
+    // At this point we know that the numbers:
+    // * have different scale
+    // * have the same sign
+    // * neither is zero
+    //
+    // The numbers have the form:
+    //
+    //      auto number = _unscaled_value * std::pow(10, -_scale);
+    //      auto number_other = other._unscaled_value * std::pow(10, -other._scale);
+    //
+    // To compare them, we make the unscaled values the same (or close), so we can directly compare the scales.
+    // To do that we want to compute diff_scale_2:
+    //
+    //      auto unscaled_diff = _unscaled_value / other._unscaled_value;
+    //      auto diff_scale_2 = log2(unscaled_diff);
+    //
+    // To avoid using division and then calculating a log2(), we use the number
+    // of bits in both numbers do infer log2(unscaled_diff) directly.
+    const int64_t bit_count = boost::multiprecision::msb(comparing_negatives ? -_unscaled_value : _unscaled_value);
```

Comment:
A thought (feel free to ignore): instead of this comparing_negatives thing, you could have swapped the two values **and** their sign - so comparing -17, -18 would change to comparing 18, 17. Then you won't need to have these comparing_negative checks.

Reply:
Also, I think the name "bit_count" is ambiguous. Wouldn't it be clearer to call it "msb" or "highest_bit"?

Reply:
Ok, sounds good.

Reply:
> We can create copies, but that can be potentially expensive

In the case when the values are negative, you already do that though. (When passing the negated argument to `msb`).

Reply:
Good point, let me see how would the code look with swap.

Reply:
Done in v6. I had to add `operator-()` too.